include/fiction/technology/sidb_cluster_hierarchy.hpp

```diff
@@ -0,0 +1,1080 @@
+#else
+    explicit sidb_cluster(phmap::flat_hash_set<sidb_ix> c, sidb_clustering v, uid_t unique_id) noexcept :
+#endif
+            uid{v.empty() ? *c.cbegin() : unique_id},
+            children{std::move(v)},
+            sidbs{std::move(c)}
+    {}
+    /**
+     * This function returns a shared pointer to the parent of this cluster.
+     *
+     * @return A shared pointer to the parent of this cluster
+     */
+    inline sidb_cluster_ptr get_parent() const
+    {
+        return parent.lock();
+    }
+    /**
+     * This function initializes the charge space of a singleton cluster corresponding with the given simulation base,
+     * and sets the initial bounds on the potential received from outside the cluster as the local potential bounds. The
+     * local potential bounds do not include the local external potential at the SiDB in the singleton, as this is
+     * passed separately.
+     *
+     * @param loc_pot_min The minimum local potential for the SiDB in the singleton cluster.
+     * @param loc_pot_max The maximum local potential for the SiDB in the singleton cluster.
+     * @param loc_ext_pot The local external potential at the SiDB in the singleton cluster. Specifically, this is the
+     * sum of the local defect potential and the local external potential.
+     * @param base The simulation base.
+     * @param self_ptr Shared pointer to itself.
+     */
+    void initialize_singleton_cluster_charge_space(const double loc_pot_min, const double loc_pot_max,
+                                                   const double loc_ext_pot, const uint8_t base,
+                                                   const sidb_cluster_ptr& self_ptr) noexcept
+    {
+        assert(sidbs.size() == 1);
+
+        const uint64_t ix = *sidbs.cbegin();
+
+        for (const sidb_charge_state cs : sidb_charge_state_iterator<sidb_state_iter_dir::TO_NEGATIVE>{
+                 base == 3 ? sidb_charge_state::POSITIVE : sidb_charge_state::NEUTRAL})
+        {
+            charge_space.emplace(sidb_cluster_charge_state{self_ptr, cs, loc_ext_pot});
+        }
+
+        recv_ext_pot_bounds[ix][static_cast<uint8_t>(bound_direction::LOWER)] = loc_pot_min;
+        recv_ext_pot_bounds[ix][static_cast<uint8_t>(bound_direction::UPPER)] = loc_pot_max;
+    }
+    /**
+     * The bound on the potential an SiDB receives from outside this cluster is returned.
+     *
+     * @tparam bound The potential bound that is requested.
+     * @param i The SiDB index for which the potential bound is requested.
+     * @return The potential bound that is stored.
+     */
+    template <bound_direction bound>
+    constexpr inline double get_recv_ext_pot_bound(const sidb_ix i) noexcept
+    {
+        return recv_ext_pot_bounds.at(i).at(static_cast<uint8_t>(bound));
+    }
+    /**
+     * The bound on the potential an SiDB receives from outside this cluster is set.
+     *
+     * @tparam bound The potential bound that is set.
```

Comment:
This is a subtle one, but I personally don't see the upper vs lower bound as types in a type theoretic sense, it's closer to an identifier really. So that makes the naming tricky.

Reply:
okay, that's a valid point!